### PR TITLE
Adding cadl sample project

### PR DESCRIPTION
**Pull request by raych1**
added a cadl sample project for DPG scenario with shared library.
referenced this sample project in the cadl-structure-guidelines document.

### CODE REVIEW

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (dw511214992) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml:5]:
```suggestion
    examples-directory: ./examples
  "@azure-tools/cadl-python":
    "sdk-folder": sdk/cntosowidgetmanager/azure-contoso-widgetmanager # delete it when you run cadl compile in local
    "package-version": "1.0.0b1"
    "package-mode": "dataplane"
    "basic-setup-py": true
    "package-name": "azure-contoso-widgetmanager"
  "@azure-tools/cadl-csharp":
    "sdk-folder": sdk/cntosowidgetmanager/Azure.Contoso.WidgetManager/src # delete it and add value true @azure-tools/cadl-csharp when you run cadl compile in local
  "@azure-tools/cadl-typescript":
    "sdk-folder": sdk/cntosowidgetmanager/widgetmanager-rest # delete it when you run cadl compile in local
    generateMetadata: true # change to false if you don't want to override meta
    generateTest: true # change to false if you don't want to generate new sample test
    azureSdkForJs: true
    packageDetails:
      name: "@azure-rest/widgetmanager-rest"
      description: "Cadl Test Service"
      version: "1.0.0-beta.1"
  "@azure-tools/cadl-java":
    "sdk-folder": sdk/cntosowidgetmanager/azure-contoso-widgetmanager # delete it when you run cadl compile in local
    namespace: com.azure.cntoso.widgetmanager
```

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/main.cadl
Comment (dw511214992) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/main.cadl:17]:
```suggestion
namespace Azure.Contoso.WidgetManager;
```

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json
Comment (dw511214992) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json:8]:
```suggestion
    "@azure-tools/cadl-azure-core": "latest",
    "@azure-tools/cadl-csharp": "latest",
    "@azure-tools/cadl-java": "latest",
    "@azure-tools/cadl-python": "latest",
    "@azure-tools/cadl-typescript": "latest",
```

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl
Comment (dw511214992) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl:1]: If there is only a change in this common file, sdk automation pipeline cannot detect to generate SDK, which is smiliar to common files in swagger, and we cannot detect which sdk needs to be regenerated.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json
Comment (weshaggard) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json:8]: We are still trying to figure out the exact plan for the emitters but I believe currently we don't want these to be in the package.json file in the specs repo but instead be in the language repos like https://github.com/Azure/azure-sdk-for-net/blob/main/eng/csharp-emitter-package.json. This is the approach that @m-nash is working on.

Also I don't believe we want this to be mapped to latest we should pin the versions, at least until things are more stable. @allenjzhang @markcowl do you guys have an opinion one how we should handle this?

On top of those concerns I think we also want to avoid having package.json files in every project but instead I think we want to have one shared one in the specs repo for defining the cadl-autorest emitter versions.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl:1]: This shared library is more likely used by this service, and SDK automation should be able identify the package under this service folder for generation.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json
Comment (dw511214992) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json:8]: @raych1  sdk automation pipeline doesn't support use package.json from SDK language repo now, and if @m-nash 's proposed solution is finalized, we can start to support it.
cc: @weshaggard

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl
Comment (dw511214992) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl:1]: Will the shared library's name always endwith `Shared`? If yes, I think we can have the enhancement: when the shared library is changed, we will regenerate all the packages.
Moreoever, I think this feature is also needed in powerapps because powerapps doesn't support multi packages in one spec PR now in current scope.

Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl:1]: Yes, that is the convention - shared library folder always ends with `Shared`.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json
Comment (weshaggard) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json:8]: >@weshaggard However, before the solution is finalized, do you need to add it to package.json in spec repo in this
sample temporilly?

If it is needed to make what we have work currently then yes I'm OK with having it here but if we don't need it yet I'd like to not have it at all from the start. Like I mentioned I'm still not sure how local cadl development will work without a local package.json file here so I'd like some feedback from the cadl team on that from @allenjzhang @markcowl

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl
Comment (allenjzhang) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl:1]: There is a standard definition of ETag in ARM. I suggest `...` spread into your custom model.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/main.cadl
Comment (allenjzhang) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/main.cadl:17]: For 1st party service, I think it starts with `Microsoft`. But you can certainly use anything and use `@armProviderNamespace` to set the Provider name used in the path.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl
Comment (allenjzhang) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl:1]: also since it is a sample project, I won't want folks to start down customization of standard models path but rather grow organically to add as needed/appropriately. So I think we can put some dummy shared/common models with instructions that if no teamwide shared/common artifacts, this project can be removed.

**Comment by dw511214992** [comment]
> ## Proposal to discuss:
> ### rest-api-spec repo
> * cadl-project.yaml
>   
>   1. autorest emitter configuration should be there. With pending new cadl config features, this should be a static file (no modification)
>   2. suggest NOT contain any SDK emitter specific configuration. these should come from SDKRepo/service folder.
> * package.json
>   
>   1. should not contain any SDK related packages info.
> 
> Repo will enforce a single set of cadl package version via "npm install". Example in step 2 below.
> 
> ### SDK repo
> * cadl-project.yaml or quevelant configuration file to store langauge specific emitter settings
> * configuration file with link to a spec folder with commit id info
> * Pipeline logic
>   
>   1. sparse-checkout folder that include all folders
>   2. with existing package.json to create a workable version by updating the package versions to a fixed SDK profile with "
>      npm install SDK-lang@0.38 SDK-@?? xxx@??
>   3. Combin/append SDK emitter config to cadl-project.yaml downloaded from rest-api-spec to get a working yaml
>   4. Build to regen code.

From my side, I have some questions after looking into the proposal:
1.	If the emitter's dependencies' versions are different from package.json’s dependencies, there will be version conflicts.
2.	For new package, there is no cadl-project.yaml in sdk folder, and then we cannot generate sdk. (The solution of swagger DPG maybe an option, will we use the solution in cadl too?)
3.	For existing package, when there is a spec PR, we cannot find which cadl-project.yaml in sdk folder is related to the changed files in the spec PR. (does the link in sdk repo’s cadl-project.yaml contains enough information to map the spec PR’s changed files to sdk repo’s cadl-project.yaml?)

**Comment by weshaggard** [comment]
Just to clarify some stuff in @allenjzhang's proposal. I do think we want to have a cadl-project.yaml file in the spec repo and it should contain general configuration such as package name and namespace for a given emitter. If there is a need for language specific configurations outside of the basic naming then those properties can come from the sdk repo. I don't think the sdk repo should have a cadl-project.yaml but instead we should have some other metadata file that will contain things like the spec version and such, something like [cadl-location.yaml](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/anomalydetector/Azure.AI.AnomalyDetector/src/cadl-location.yaml) (we could rename it if necessary). In that file we could add additional properties/config if they are needed for the emitter. 


To help answer @dw511214992's questions:
1) The goal will be to not have any emitter versions in the package.json file in the specs repo. If there are versions in there we should either update them to be consistent. We might need to update all the package versions for every package dependency of the emitter we are trying to run.
2) For new sdk packages we should generate a basic cadl-location.yaml file (not cadl-project.yaml), see my earlier comment. If one doesn't exist then we should generate one from the template just like we will have to generate a project as well.
3) Yes I would propose that this be the cadl-location.yaml file and it should have the pointers we need to where the spec is contained and it should be updated when we are updating the spec.
4) Good question that I didn't consider yet. I think the list of emitters should still remain in the cald-project.yaml file in the specs repo. This as well as naming configuration is a good reason to keep them in there.

**Comment by allenjzhang** [comment]
> Just to clarify some stuff in @allenjzhang's proposal. I do think we want to have a cadl-project.yaml file in the spec repo and it should contain general configuration such as package name and namespace for a given emitter. If there is a need for language specific configurations outside of the basic naming then those properties can come from the sdk repo. I don't think the sdk repo should have a cadl-project.yaml but instead we should have some other metadata file that will contain things like the spec version and such, something like [cadl-location.yaml](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/anomalydetector/Azure.AI.AnomalyDetector/src/cadl-location.yaml) (we could rename it if necessary). In that file we could add additional properties/config if they are needed for the emitter. 
> 
> 
**Yup**


> To help answer @dw511214992's questions:
> 
> 1) The goal will be to not have any emitter versions in the package.json file in the specs repo. If there are versions in there we should either update them to be consistent. We might need to update all the package versions for every package dependency of the emitter we are trying to run.
> 
> 2) For new sdk packages we should generate a basic cadl-location.yaml file (not cadl-project.yaml), see my earlier comment. If one doesn't exist then we should generate one from the template just like we will have to generate a project as well.
> 
> 
1, 2 Agreed.

**Comment by allenjzhang** [comment]
Regarding #4, my think is spec repo will a only contain Cadl-autorest emitter entry. For each language SDK repo, the emitter name will be injected either from command line or manipulation result of Cadl-project.yaml. Most likely the former. 

For spec ci pipeline where everything need to be generated, I assume all lang emitter will be passed on CLI.

**Comment by weshaggard** [comment]
>Regarding https://github.com/Azure/azure-rest-api-specs/pull/4, my think is spec repo will a only contain Cadl-autorest emitter entry. For each language SDK repo, the emitter name will be injected either from command line or manipulation result of Cadl-project.yaml. Most likely the former.

For spec ci pipeline where everything need to be generated, I assume all lang emitter will be passed on CLI.

I agree the project.json should only contain the swagger emitter entry but I think we need all the supported emitters and their configurations listed in the cadl-project.yaml file. They will need configurations and they should be listed on each emitter specifically.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/package.json
Comment (weshaggard) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/package.json:10]: This should be reduced to `@azure-tools/cadl-autorest": "latest",`. We also should confirm with the cadl folks whether this should be under dependencies or devDependencies? @allenjzhang does it matter from a cadl perspective?

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (weshaggard) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml:5]: As we chatted I think all of this makes sense except for the versions we should not have package version info in this set of configurations.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml:18]: Will use interpolated path once the feature is available.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.cadl]: Updated to use a faked shared model.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/package.json]: Updated.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/main.cadl
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/main.cadl]: Use 'Azure' to be consistent with generated sdk namespace.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml]: Updated.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/package.json
Comment (weshaggard) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/package.json:1]: Do we need have a package.json file for the shared library?

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (weshaggard) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml:19]: This should use the `{openapi-folder}`

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (weshaggard) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml:11]:
```suggestion
    default: "{project-root}/../data-plane/"
```

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [inline documentation/cadl-structure-guidelines.md:1]: We should update this doc and add pointers to the new docs being worked on at https://github.com/Azure/azure-rest-api-specs/pull/21722. We should also make sure those docs match these changes in this PR.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/data-plane/Contoso.WidgetManager/preview/2022-11-01-preview/examples/Widgets_DeleteWidgetSample.json
Comment (lmazuel) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/data-plane/Contoso.WidgetManager/preview/2022-11-01-preview/examples/Widgets_DeleteWidgetSample.json:1]: There should not be a data-plane folder

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/main.cadl
Comment (lmazuel) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/main.cadl:3]: I don't think this file is using the autorest extension at all. We should not import thing we don't use, especially autorest should be limited.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml:19]: `openapi-folder` isn't supported currently.
https://github.com/Azure/cadl-azure/blob/main/packages/cadl-autorest/src/lib.ts#L12

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/package.json
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/package.json]: Removed package.json for shared library.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/package.json
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager.Shared/package.json]: Removed.

**Comment by m-nash** [comment]
> > Regarding #4, my think is spec repo will a only contain Cadl-autorest emitter entry. For each language SDK repo, the emitter name will be injected either from command line or manipulation result of Cadl-project.yaml. Most likely the former.
> 
> For spec ci pipeline where everything need to be generated, I assume all lang emitter will be passed on CLI.
> 
> I agree the project.json should only contain the swagger emitter entry but I think we need all the supported emitters and their configurations listed in the cadl-project.yaml file. They will need configurations and they should be listed on each emitter specifically.

I also lean towards keeping all emitter configurations in the cadl-project.yaml in the spec repo and not having one in the sdk folder.  There are no technical blockers from doing this that I am aware of right now and it fits with our long term goals of being able to fully generate out of the spec repo (even if its just a script that wraps generating from all the language repos).

**Review by weshaggard — APPROVED** [review]
I believe this matches all the conversations I've been involved in. We should at least get @allenjzhang to also sign-off on this but then I think we should move forward with this, and merge and we can always update if we find issues.

**Review by dw511214992 — APPROVED** [review]

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (allenjzhang) [inline documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml:17]: Would love to see this use a token instead of relative path ../../../..

**Comment by allenjzhang** [comment]
Depending on how you want to use the samples here as a demonstrator or a receipt for a user to follow. If later, we will also need to introduce two different flavors Data-plane and ARM to ensure a good out-of-box user experience, where DP should include cadl-dpg in the package.config. 

My thinking is the former. It's a demostration project and for us to validate and test. In the official getting started step, we will ask the user to use `cadl init azure` to select from a list of templates that we can refine later without holding up this PR.

File: documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml
Comment (raych1) [reply documentation/cadl-sample-project/specification/contosowidgetmanager/Contoso.WidgetManager/cadl-project.yaml:17]: It's confusing but this is correct relative path base on cadl compiler@0.38.5. We could update it when cadl compiler fix it.

This sample project is for data plane scenario so the value will be data-plane only.

**Comment by raych1** [comment]
> Depending on how you want to use the samples here as a demonstrator or a receipt for a user to follow. If later, we will also need to introduce two different flavors Data-plane and ARM to ensure a good out-of-box user experience, where:
> 
> * DP should include cadl-dpg in the package.config
> * include generate model flag true by default,
> * having a separate client.cadl file with some basic client configurations.
> 
> My thinking is the former. It's a demostration project and for us to validate and test. In the official getting started step, we will ask the user to use `cadl init azure` to select from a list of templates that we can refine later without holding up this PR.

Agree, we can add data plane and management plane templates later to `cadl init azure` later.

**Pull request merged**
